pkg/resourcemanager/webhook/projectedtokenmount/handler.go

```diff
@@ -0,0 +1,152 @@
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/utils/pointer"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+)
+
+type handler struct {
+	targetClient      client.Client
+	decoder           *admission.Decoder
+	expirationSeconds int64
+}
+
+// NewHandler returns a new handler.
+func NewHandler(targetClient client.Client, expirationSeconds int64) admission.Handler {
+	return &handler{
+		targetClient:      targetClient,
+		expirationSeconds: expirationSeconds,
+	}
+}
+
+func (h *handler) InjectDecoder(d *admission.Decoder) error {
+	h.decoder = d
+	return nil
+}
+
+func (h *handler) Handle(ctx context.Context, req admission.Request) admission.Response {
+	pod := &corev1.Pod{}
+	if err := h.decoder.Decode(req, pod); err != nil {
+		return admission.Errored(http.StatusUnprocessableEntity, err)
+	}
+
+	if metav1.HasAnnotation(pod.ObjectMeta, resourcesv1alpha1.ProjectedTokenSkip) {
+		return admission.Allowed("pod explicitly opts out of projected service account token mount")
+	}
+
+	if len(pod.Spec.ServiceAccountName) == 0 || pod.Spec.ServiceAccountName == "default" {
+		return admission.Allowed("service account not specified or defaulted")
+	}
+
+	serviceAccount := &corev1.ServiceAccount{}
+	if err := h.targetClient.Get(ctx, kutil.Key(pod.Namespace, pod.Spec.ServiceAccountName), serviceAccount); err != nil {
+		return admission.Errored(http.StatusInternalServerError, err)
+	}
+
+	if serviceAccount.AutomountServiceAccountToken == nil || *serviceAccount.AutomountServiceAccountToken {
+		return admission.Allowed("auto-mounting service account token is not disabled")
+	}
+
+	for _, volume := range pod.Spec.Volumes {
+		if strings.HasPrefix(volume.Name, serviceAccountVolumeNamePrefix) {
+			return admission.Allowed("pod already has service account defaultVolume mount")
+		}
+	}
+
+	pod.Spec.Volumes = append(pod.Spec.Volumes, getVolume(h.expirationSeconds))
+	for i := range pod.Spec.Containers {
+		pod.Spec.Containers[i].VolumeMounts = append(pod.Spec.Containers[i].VolumeMounts, getVolumeMount())
+	}
+
+	marshaledPod, err := json.Marshal(pod)
+	if err != nil {
+		return admission.Errored(http.StatusInternalServerError, err)
+	}
+
+	return admission.PatchResponseFromRaw(req.Object.Raw, marshaledPod)
+}
+
+const (
+	serviceAccountVolumeNamePrefix = "kube-api-access-"
+	serviceAccountVolumeNameSuffix = "gardener"
+)
+
+func volumeName() string {
+	return serviceAccountVolumeNamePrefix + serviceAccountVolumeNameSuffix
+}
+
+func getVolume(expirationSeconds int64) corev1.Volume {
+	return corev1.Volume{
+		Name: volumeName(),
+		VolumeSource: corev1.VolumeSource{
+			Projected: &corev1.ProjectedVolumeSource{
+				DefaultMode: pointer.Int32(420),
+				Sources: []corev1.VolumeProjection{
+					{
+						ServiceAccountToken: &corev1.ServiceAccountTokenProjection{
+							ExpirationSeconds: &expirationSeconds,
+							Path:              "token",
```

Comment:
Yes, I agree, I don't think it would provide additional value. Apart from that, since the webhook would also be enabled for the seeds, using audiences wouldn't be easily possible anyways (w/o additional complexity) as we don't know which audiences are valid for a seed cluster.